[SPARK-7810] [pyspark] solve python rdd socket connection problem

aihex opened the pull request:
Method "_load_from_socket" in rdd.py cannot load data from jvm socket when ipv6 is used. The current method only works well with ipv4. New modification should work around both two protocols.

Commit:
solve python rdd socket connection problem to jvm

AmplabJenkins commented:
Can one of the admins verify this patch?

JoshRosen commented:
Jenkins, this is ok to test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33776 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33776/consoleFull) for   PR 6338 at commit [`5644953`](https://github.com/apache/spark/commit/564495305d1a8af4e18d3a3b220431645ea0b296).

SparkQA commented:
[Test build #33776 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33776/consoleFull) for   PR 6338 at commit [`5644953`](https://github.com/apache/spark/commit/564495305d1a8af4e18d3a3b220431645ea0b296).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AiHe commented:
@JoshRosen Could you let me know how to figure out the reason for this failure?

SparkQA commented:
[Test build #935 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/935/consoleFull) for   PR 6338 at commit [`5644953`](https://github.com/apache/spark/commit/564495305d1a8af4e18d3a3b220431645ea0b296).

SparkQA commented:
[Test build #935 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/935/console) for   PR 6338 at commit [`5644953`](https://github.com/apache/spark/commit/564495305d1a8af4e18d3a3b220431645ea0b296).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AiHe commented:
@JoshRosen Hi, can you kindly review this PR again?

JoshRosen commented on python/pyspark/rdd.py:
```diff
@@ -121,14 +121,21 @@ def _parse_memory(s):
+            sock = None
+            continue
+        break
+    if sock:
```
If `sock is None` (the logical `else` branch), should we log an error or throw an exception?

JoshRosen commented on python/pyspark/rdd.py:
```diff
@@ -121,14 +121,21 @@ def _parse_memory(s):
         rf = sock.makefile("rb", 65536)
         for item in serializer.load_stream(rf):
             yield item
-    finally:
```
I think that we'll still want to keep a `sock.close()` call in a `finally` block to ensure that the socket is promptly closed if a deserialization error occurs.

JoshRosen replied:
I think we can do this by just indenting the contents of this `if` branch and putting them into a `try-finally`.

JoshRosen commented on python/pyspark/rdd.py:
```diff
@@ -121,14 +121,21 @@ def _parse_memory(s):
-    try:
-        sock.connect(("localhost", port))
+    sock = None
+    for res in socket.getaddrinfo("localhost", port, socket.AF_UNSPEC, socket.SOCK_STREAM):
```
Do you mind adding a one- or two-line comment to explain why we're looping over these interfaces?

JoshRosen commented:
Thanks for the fix and sorry to lag behind in reviewing this; I've been super swamped with other stuff, so thanks for pinging me as a reminder.

This seems like a good change overall; my only comments relate to a possible error-handling corner case: I want to make sure that we don't accidentally forget to close the sockets if deserialization errors occur.

Commit:
handle code review

1. explain loop

2. add try-finally block to close socket when there exists connection error

3. throw exception when socket connection is not established correctly

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35588/consoleFull) for   PR 6338 at commit [`e75c5c8`](https://github.com/apache/spark/commit/e75c5c8e7039e1658cbd2e6e12f404f133131ede).

SparkQA commented:
[Test build #35588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35588/console) for   PR 6338 at commit [`e75c5c8`](https://github.com/apache/spark/commit/e75c5c8e7039e1658cbd2e6e12f404f133131ede).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AiHe commented:
It's confusing that all test have been passed but it ends up with a failure.

JoshRosen commented:
Jenkins, retest this please.

JoshRosen commented:
The error message from AMPLab Jenkins is due to a problem in our Jenkins configuration, not a problem in tests. Let me run this again and I'll merge once SparkQA says that it's okay.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35995/consoleFull) for   PR 6338 at commit [`e75c5c8`](https://github.com/apache/spark/commit/e75c5c8e7039e1658cbd2e6e12f404f133131ede).

SparkQA commented:
[Test build #35995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35995/console) for   PR 6338 at commit [`e75c5c8`](https://github.com/apache/spark/commit/e75c5c8e7039e1658cbd2e6e12f404f133131ede).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Actually, I have one concern about exception masking; will comment inline.

JoshRosen commented on python/pyspark/rdd.py:
```diff
@@ -121,15 +121,30 @@ def _parse_memory(s):
+            for item in serializer.load_stream(rf):
+                yield item
+        except socket.error:
+            raise Exception("encounter error when connecting to socket server")
```
Because Python 2.x doesn't support chained exceptions, I think this will end up masking the actual cause of the socket error, which could make errors here really hard to debug.  Can we just omit the `except` block and allow the original exception to propagate instead?

AiHe replied:
Yes, you are right. The situation here is not the same as the one above when socket.connect().

davies replied:
+1 for not having the `except socket.error`, or you need to put the underline exception in the message.

JoshRosen commented:
Also, /cc @davies for another pair of eyes.  This seems to match up with the example at https://docs.python.org/2/library/socket.html#example, so it seems okay, but I just want someone else to take a look, too.

AiHe commented:
When I'm working on python networking problem, I usually follow that example as a python standard usage. Then I just apply the underlying idea here. I'm wondering if I need some reference work although it's just an example of the language's documentation.

davies commented on python/pyspark/rdd.py:
```diff
@@ -121,15 +121,30 @@ def _parse_memory(s):
+        finally:
+            sock.close()
+    else:
+        raise Exception("could not open socket")
```
I'd like put this branch near the connecting part

AiHe replied:
Hi Davies, you mean the "else branch"? I think it will throw the exception once connections in both IPV4 and IPV6 cannot be established. What's the benefit to move the "else" branch?

davies replied:
It's not about correctness, it will make the code easy to read.

Commit:
handle code review 2

1. remove exception of socket.error for keeping tracking of error outside

2. refactor for better readability

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36038/consoleFull) for   PR 6338 at commit [`d4fc9c4`](https://github.com/apache/spark/commit/d4fc9c4d4332f004d429b69b514345f30f026d58).

SparkQA commented:
[Test build #36038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36038/console) for   PR 6338 at commit [`d4fc9c4`](https://github.com/apache/spark/commit/d4fc9c4d4332f004d429b69b514345f30f026d58).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

davies commented:
LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

davies commented:
Merged into master, 1.4 and 1.3 branch

AiHe commented:
Thank you guys!
